kat.go

```diff
@@ -0,0 +1,326 @@
+// Copyright 2023 Contributors to the Veraison project.
+// SPDX-License-Identifier: Apache-2.0
+package parsectpm
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rsa"
+	"fmt"
+
+	tpm2 "github.com/google/go-tpm/tpm2"
+	"github.com/google/go-tpm/tpmutil"
+	"github.com/veraison/eat"
+	cose "github.com/veraison/go-cose"
+)
+
+type KAT struct {
+	TpmVer   *string `cbor:"tpmVer" json:"tpmVer"`
+	KID      *[]byte `cbor:"kid" json:"kid"`
+	Sig      *[]byte `cbor:"sig" json:"sig"`
+	PubArea  *[]byte `cbor:"pubArea" json:"pubArea"`
+	CertInfo *[]byte `cbor:"certInfo" json:"certInfo"`
+}
+
+func NewKAT() *KAT {
+	return &KAT{}
+}
+
+func (k *KAT) SetTpmVer(v string) error {
+	if v == "" {
+		return fmt.Errorf("empty string specified")
+	}
+	k.TpmVer = &v
+	return nil
+}
+
+func (k *KAT) SetKeyID(v []byte) error {
+	data := eat.UEID(v)
+	if err := data.Validate(); err != nil {
+		return fmt.Errorf("failed to validate UEID: %w", err)
+	}
+
+	k.KID = &v
+	return nil
+}
+
+func (k *KAT) SetSig(s []byte) error {
+	k.Sig = &s
+	return nil
+}
+
+func (k KAT) Validate() error {
+	if k.TpmVer == nil {
+		return fmt.Errorf("TPM Version not set")
+	} else if *k.TpmVer == "" {
+		return fmt.Errorf("Empty TPM Version")
+	}
+
+	if k.KID == nil {
+		return fmt.Errorf("missing key identifier")
+	}
+	if err := validateKID(*k.KID); err != nil {
+		return fmt.Errorf("invalid KID : %w", err)
+	}
+
+	if k.CertInfo == nil {
+		return fmt.Errorf("no certificate information")
+	}
+	_, err := tpm2.DecodeAttestationData(*k.CertInfo)
+	if err != nil {
+		return fmt.Errorf("failed to decode supplied certification information %w", err)
+	}
+
+	if k.Sig == nil {
+		return fmt.Errorf("missing signature")
+	}
+	// Check the signature decode results in a success or not?
+	_, err = tpm2.DecodeSignature(bytes.NewBuffer(*k.Sig))
+	if err != nil {
+		return fmt.Errorf("not a valid signature")
+	}
+
+	if k.PubArea == nil {
+		return fmt.Errorf("missing public key information")
+	}
+	_, err = tpm2.DecodePublic(*k.PubArea)
+	if err != nil {
+		return fmt.Errorf("unable to decode the Public Area %w", err)
+	}
+	return nil
+}
+
+type DigestInfo struct {
+	HashAlgID uint64
+	Digest    []byte
+}
+
+type NameInfo struct {
+	Handle  uint32
+	DigInfo DigestInfo
+}
+
+type TpmCertInfo struct {
+	Name          NameInfo
+	QualifiedName NameInfo
+}
+
+type CertInfo struct {
+	Magic       uint32
+	Type        uint16
```

Comment:
if magic and type (and any other spec-defined MUST) have predefined values, they need to be enforced in this package.  We can't expect the caller to know the spec.